realm/realm-library/src/main/cpp/io_realm_SyncSession.cpp

```diff
@@ -49,3 +54,39 @@ JNIEXPORT jboolean JNICALL Java_io_realm_SyncSession_nativeRefreshAccessToken(JN
     CATCH_STD()
     return JNI_FALSE;
 }
+
+JNIEXPORT jboolean JNICALL Java_io_realm_SyncSession_nativeWaitForDownloadCompletion(JNIEnv* env,
+                                                                                     jobject session_object,
+                                                                                     jstring localRealmPath)
+{
+    TR_ENTER()
+    try {
+        JStringAccessor local_realm_path(env, localRealmPath);
+        auto session = SyncManager::shared().get_existing_session(local_realm_path);
+
+        if (session) {
+            static jni_util::JavaClass java_sync_session_class(env, "io/realm/SyncSession");
+            static jni_util::JavaMethod java_notify_result_method(
+                env, java_sync_session_class, "notifyAllChangesDownloaded", "(Ljava/lang/Long;Ljava/lang/String;)V");
+            jni_util::JavaGlobalRef java_session_object_ref(env, session_object);
+
+            bool listener_registered =
+                session->wait_for_download_completion([java_session_object_ref](std::error_code error) {
+                    JNIEnv* env = realm::jni_util::JniUtils::get_env(true);
+                    jni_util::JavaLocalRef<jobject> java_error_code_ref(env, nullptr);
+                    jni_util::JavaLocalRef<jstring> java_error_message_ref(env, nullptr);
+                    if (error != std::error_code{}) {
+                        java_error_code_ref = jni_util::JavaLocalRef<jobject>(env, NewLong(env, error.value()));
+                        java_error_message_ref =
+                            jni_util::JavaLocalRef<jstring>(env, env->NewStringUTF(error.message().c_str()));
```

Comment:
uh, when construct `StringData` from `std::string`, a `std::string` instance needs to be retained to make sure the data is valid. So `NewStringUTF` is fine here. just ignore me.